src/raftstore/store/snap.rs

```diff
@@ -618,6 +620,987 @@ mod v1 {
+                cf_file.file = Some(f);
+            }
+            let f = try!(OpenOptions::new()
+                .write(true)
+                .create_new(true)
+                .open(&self.meta_file.tmp_path));
+            self.meta_file.file = Some(f);
+            Ok(())
+        }
+
+        fn init_for_applying(&mut self) -> RaftStoreResult<()> {
+            self.load_snapshot_metadata()
+        }
+
+        fn load_snapshot_metadata(&mut self) -> RaftStoreResult<()> {
+            let size = try!(get_file_size(&self.meta_file.path));
+            let mut file = try!(File::open(&self.meta_file.path));
+            let mut buf = Vec::with_capacity(size as usize);
+            try!(file.read_to_end(&mut buf));
+            let mut snapshot_data = RaftSnapshotData::new();
+            try!(snapshot_data.merge_from_bytes(&buf));
+            self.meta_file.data = snapshot_data;
+            Ok(())
+        }
+
+        fn get_display_path(dir_path: &PathBuf, prefix: &str) -> String {
+            let mut cf_names = String::from("");
+            for (i, cf) in get_snapshot_cfs().iter().enumerate() {
+                if i == 0 {
+                    cf_names += "(";
+                    cf_names += cf;
+                } else {
+                    cf_names += "|";
+                    cf_names += cf;
+                }
+            }
+            cf_names += ")";
+            format!("{}/{}_{}{}",
+                    dir_path.as_path().to_str().unwrap().to_owned(),
+                    prefix,
+                    cf_names,
+                    SST_FILE_SUFFIX)
+        }
+
+        fn try_delete(&self) -> io::Result<()> {
+            debug!("deleting {}", self.path());
+            let exists = self.exists();
+            let mut total_size = 0;
+            for cf_file in &self.cf_files {
+                if file_exists(&cf_file.tmp_path) {
+                    delete_file(&cf_file.tmp_path);
+                }
+                if exists {
+                    total_size += try!(get_file_size(&cf_file.path));
+                }
+                delete_file(&cf_file.path);
+            }
+            delete_file(&self.meta_file.tmp_path);
+            delete_file(&self.meta_file.path);
+            let size_to_sub = if exists {
+                total_size
+            } else if self.meta_file.data.has_file_size() {
+                self.meta_file.data.get_file_size()
+            } else {
+                0
+            };
+            if size_to_sub != 0 {
+                let mut size_track = self.size_track.wl();
+                *size_track = size_track.saturating_sub(size_to_sub);
+            }
+            Ok(())
+        }
+
+        fn validate(&self) -> RaftStoreResult<()> {
+            let cf_size_checksums =
+                box_try!(decode_cf_size_checksums(self.meta_file.data.get_data()));
+            for (cf, expected_size, expected_checksum) in cf_size_checksums {
+                for cf_file in &self.cf_files {
```

Comment:
I prefer 
```
for cf_file in &self.cf_files {
    find size and checksum from cf_size_checksums, or validate failed
}
```
this can guarantee that all cfs have validated.